# Lab-1: Data Manipulation and Exploration.
_____
##### In this lab, you will manipulate a dataset and prepare it for training and testing.

### Objectives:
1. Data exploration
2. One-hot-encoding of categorical features
3. Data imputation
4. Features Scaling
5. Trainset splitting
6. Data plotting
-----
### Loading Data

In [1]:
import pandas as pd

data = pd.read_csv('./data/train.csv')
x_test_private = pd.read_csv('./data/test.csv')
x_train = data.loc[:, :'Gender Inequality Index (GII)']
y_train = data['y']

In [2]:
data.head(5)

galactic year                        galaxy  existence expectancy index  \
0         990025  Large Magellanic Cloud (LMC)                    0.628657   
1         990025              Camelopardalis B                    0.818082   
2         990025                       Virgo I                    0.659443   
3         990025            UGC 8651 (DDO 181)                    0.555862   
4         990025                  Tucana Dwarf                    0.991196   

   existence expectancy at birth  Gross income per capita  Income Index  \
0                      63.125200             27109.234310      0.646039   
1                      81.004994             30166.793958      0.852246   
2                      59.570534              8441.707353      0.499762   
3                      52.333293                      NaN           NaN   
4                      81.802464             81033.956906      1.131163   

   Expected years of education (galactic years)  \
0                                      8.240543   
1                                     10.671823   
2                                      8.840316   
3                                           NaN   
4                                     13.800672   

   Mean years of education (galactic years)  \
0                                       NaN   
1                                  4.742470   
2                                  5.583973   
3                                       NaN   
4                                 13.188907   

   Intergalactic Development Index (IDI)  Education Index  ...  \
0                                    NaN              NaN  ...   
1                               0.833624         0.467873  ...   
2                               0.469110         0.363837  ...   
3                                    NaN              NaN  ...   
4                               0.910341         0.918353  ...   

   Intergalactic Development Index (IDI), female  \
0                                            NaN   
1                                            NaN   
2                                            NaN   
3                                            NaN   
4                                            NaN   

   Intergalactic Development Index (IDI), male  \
0                                          NaN   
1                                          NaN   
2                                          NaN   
3                                          NaN   
4                                          NaN   

   Gender Development Index (GDI)  \
0                             NaN   
1                             NaN   
2                             NaN   
3                             NaN   
4                             NaN   

   Intergalactic Development Index (IDI), female, Rank  \
0                                                NaN     
1                                                NaN     
2                                                NaN     
3                                                NaN     
4                                                NaN     

   Intergalactic Development Index (IDI), male, Rank  Adjusted net savings   \
0                                                NaN                    NaN   
1                                                NaN              19.177926   
2                                                NaN              21.151265   
3                                                NaN                    NaN   
4                                                NaN                    NaN   

   Creature Immunodeficiency Disease prevalence, adult (% ages 15-49), total  \
0                                                NaN                           
1                                                NaN                           
2                                           6.534020                           
3                                           5.912194                           
4                                  

### Data exploration

In [3]:
x_train.head(5)

galactic year                        galaxy  existence expectancy index  \
0         990025  Large Magellanic Cloud (LMC)                    0.628657   
1         990025              Camelopardalis B                    0.818082   
2         990025                       Virgo I                    0.659443   
3         990025            UGC 8651 (DDO 181)                    0.555862   
4         990025                  Tucana Dwarf                    0.991196   

   existence expectancy at birth  Gross income per capita  Income Index  \
0                      63.125200             27109.234310      0.646039   
1                      81.004994             30166.793958      0.852246   
2                      59.570534              8441.707353      0.499762   
3                      52.333293                      NaN           NaN   
4                      81.802464             81033.956906      1.131163   

   Expected years of education (galactic years)  \
0                                      8.240543   
1                                     10.671823   
2                                      8.840316   
3                                           NaN   
4                                     13.800672   

   Mean years of education (galactic years)  \
0                                       NaN   
1                                  4.742470   
2                                  5.583973   
3                                       NaN   
4                                 13.188907   

   Intergalactic Development Index (IDI)  Education Index  ...  \
0                                    NaN              NaN  ...   
1                               0.833624         0.467873  ...   
2                               0.469110         0.363837  ...   
3                                    NaN              NaN  ...   
4                               0.910341         0.918353  ...   

   Current health expenditure (% of GGP)  \
0                                    NaN   
1                                    NaN   
2                                    NaN   
3                                    NaN   
4                                    NaN   

   Intergalactic Development Index (IDI), female  \
0                                            NaN   
1                                            NaN   
2                                            NaN   
3                                            NaN   
4                                            NaN   

   Intergalactic Development Index (IDI), male  \
0                                          NaN   
1                                          NaN   
2                                          NaN   
3                                          NaN   
4                                          NaN   

   Gender Development Index (GDI)  \
0                             NaN   
1                             NaN   
2                             NaN   
3                             NaN   
4                             NaN   

   Intergalactic Development Index (IDI), female, Rank  \
0                                                NaN     
1                                                NaN     
2                                                NaN     
3                                                NaN     
4                                                NaN     

   Intergalactic Development Index (IDI), male, Rank  Adjusted net savings   \
0                                                NaN                    NaN   
1                                                NaN              19.177926   
2                                                NaN              21.151265   
3                                                NaN                    NaN   
4                                                NaN                    NaN   

   Creature Immunodeficiency Disease prevalence, adult (% ages 15-49), total  \
0                                                NaN                           
1         

In [4]:
types = x_train.dtypes
print("Number categorical featues:", sum(types=='object'))
print(types)

Number categorical featues: 1
galactic year                                                                  int64
galaxy                                                                        object
existence expectancy index                                                   float64
existence expectancy at birth                                                float64
Gross income per capita                                                      float64
                                                                              ...   
Intergalactic Development Index (IDI), male, Rank                            float64
Adjusted net savings                                                         float64
Creature Immunodeficiency Disease prevalence, adult (% ages 15-49), total    float64
Private galaxy capital flows (% of GGP)                                      float64
Gender Inequality Index (GII)                                                float64
Length: 79, dtype: object


In [5]:
x_train.describe()

galactic year  existence expectancy index  \
count   3.865000e+03                 3864.000000   
mean    1.000709e+06                    0.872479   
std     6.945463e+03                    0.162367   
min     9.900250e+05                    0.227890   
25%     9.950060e+05                    0.763027   
50%     1.000000e+06                    0.907359   
75%     1.006009e+06                    0.992760   
max     1.015056e+06                    1.246908   

       existence expectancy at birth  Gross income per capita  Income Index  \
count                    3864.000000              3837.000000   3837.000000   
mean                       76.798111             31633.240872      0.825154   
std                        10.461654             18736.378445      0.194055   
min                        34.244062              -126.906522      0.292001   
25%                        69.961449             20169.118912      0.677131   
50%                        78.995101             26600.768195      0.827300   
75%                        84.558971             36898.631754      0.970295   
max                       100.210053            151072.683156      1.361883   

       Expected years of education (galactic years)  \
count                                   3732.000000   
mean                                      14.723296   
std                                        3.612546   
min                                        3.799663   
25%                                       12.592467   
50%                                       14.942913   
75%                                       17.123797   
max                                       26.955944   

       Mean years of education (galactic years)  \
count                               3502.000000   
mean                                  10.283959   
std                                    3.319948   
min                                    1.928166   
25%                                    7.654169   
50%                                   10.385465   
75%                                   12.884752   
max                                   19.057648   

       Intergalactic Development Index (IDI)  Education Index  \
count                            3474.000000      3474.000000   
mean                                0.804246         0.745900   
std                                 0.176242         0.199795   
min                                 0.273684         0.189874   
25%                                 0.671862         0.597746   
50%                                 0.824758         0.761255   
75%                                 0.939043         0.893505   
max                                 1.232814         1.269625   

       Intergalactic Development Index (IDI), Rank  ...  \
count                                  3432.000000  ...   
mean                                    135.129178  ...   
std                                      52.449535  ...   
min                                       9.925906  ...   
25%                                      92.262724  ...   
50%                                     135.914318  ...   
75%                                     175.301993  ...   
max                                     278.786613  ...   

       Current health expenditure (% of GGP)  \
count                             867.000000   
mean                                8.837096   
std                                 2.806800   
min                                 2.313778   
25%                                 6.843091   
50%                                 8.518629   
75%                                10.573365   
max                                23.531546   

       Intergalactic Development Index (IDI), female  \
count                                     916.000000   
mean                                        0.823561   
std                                         0.185780   
min                                         0.305733   
25%                                    

### One-hot-encoding of categorical feature
1. Why to encode the categorical feature?
<span style="color:blue">Because most of the machine learning and basic deep learning algorithms can only accept numerical data and features.</span>
2. Why one-hot-encoding form? why not give each category in the feature a specific code value?
<span style="color:blue"> Giving each category a number (e.g. giving categories: red, blue, black -> 1, 2, 3) will indicate to the learning algorithm that there's some ordering (e.g., black is larger than red) which is not true as the categories are nominal.
3. How many new features will be added? What if we have some missing values, which is the case in this dataset?
<span style="color:blue"> Usually the number of new features is #categories - 1, as one indicator variable can be infered from the others (e.g., if not black and blue then it is red for sure, so black -> [1, 0], blue -> [0, 1], red -> [0, 0]). But if the trainset may have empty values for this feature, and there's a posibility that the testset also has empty cells for this feature, then it is better to have #categories new features to indicate missing value (i.e., black -> [1, 0, 0], blue -> [0, 1, 0], red -> [0, 0, 1], NA -> [0, 0, 0]).

4. When is it appropriate to give each category an encoding value?
<span style="color:blue"> It will be appropriate if the categorical feature is ordinal (e.g, small, medium, large). Then it is reasonable approach to give them ordered values without transform it to one-hot-encoding form (i.e., small -> 1, medium -> 2, large -> 3)
##### In the next cell, implement the function that take a dataframe, name of the categorical feature, and the encoder object. Then, adds new features that represent the one-hot-encoding form of this feature and ignore the missing values in it (encode them to zeros).
##### Then encode the same column/feature in the testset with the same encoder object.

In [6]:
from sklearn.preprocessing import OneHotEncoder, PolynomialFeatures

encoder = OneHotEncoder(handle_unknown='ignore', sparse=False)
cat_feats = ['galaxy']
encoder.fit(x_train[cat_feats])

def ohe_new_features(df, features_name, encoder):
    new_feats = encoder.transform(df[features_name])
    new_cols = pd.DataFrame(new_feats, dtype=int)
    new_df = pd.concat([df, new_cols], axis=1)
    new_df.drop(features_name, axis=1, inplace=True)
    return new_df


# Write your code here (2 lines)
# Transform both train and testset.
x_train = ohe_new_features(x_train, cat_feats, encoder)
x_test_private = ohe_new_features(x_test_private, cat_feats, encoder)

### Data Imputation
#### Imputation is filling the missing values in the dataset.
Several issues to address:
1. Why we should fill them? Why not remove the rows that has missing values?
<span style="color:blue"> Most of the machine learning algorithms can't deal with missing values cell (NA/NAN). Removing the rows/cols that have missing values could be an option when the missing values in some rows/cols is a few and the to-be-deleted rows/cols are few comparing to what's left (removing 50 rows out of 10k rows that have 50 empty values on average for 80 features). However, the usual case it not be deleted (why?), Some information will be lost due to the deletion which is the non-missing values, so it is better not to lose them.
2. Filling it with constant value for each value, but what's it?
<span style="color:blue"> Estimating a constant value that best to represent the feature is trivially chosen to be the expected value of this feature, which is the mean. Other options is to use some other central tendency measures like the median, mode or mid-range but the best practice is to use mean or median for numerical continuous features and use mode for the discrete ones.
3. (Advanced) Estimate the filling value, each row for its own.
<span style="color:blue"> There's another option of filling the missing values is to infer them from the rest of the data by regression/classification models (one for each feature).
This method is out of the level of this course so you can read about it [here](https://scikit-learn.org/stable/modules/generated/sklearn.impute.IterativeImputer.html#sklearn.impute.IterativeImputer) and [here](https://www.kaggle.com/shashankasubrahmanya/missing-data-imputation-using-regression).
4. How to choose the imputation strategy?
<span style="color:blue"> It can be treated as a hyperparamter and choose the best option using a validation set or cross-validation (to be explained).
4. (Not related only to imputations) Why we fit the imputer to the training data not the concatenation of both train and test set?
<span style="color:blue"> This would cause data leakage to the model. This means the model got some information from the testset which shouldn't be considered at all before the testing step. So, the testset shouldn't affect the preprocessing in anyway (e.g., data imputation or scaling).
Same goes for validation set.

##### In the next cell, implement a function that returns the count of nan/empty cells in a dataframe.
##### Use SimpleImputer object, fit it to the trainset, then transform both the train and test sets.

In [7]:
from sklearn.impute import SimpleImputer
import numpy as np

# Complete a function that returns number of empty cells in
def count_nans(df):
    # 1 or 2 lines.
    return pd.isna(df).sum().sum()
# Print number of empty cells both in x_train and x_test_private (1 line)
print("#Empty cells in x_train=", count_nans(x_train))
print("#Empty cells in x_test_private=", count_nans(x_test_private))

# Write your code here (4 lines):
# create the imputer object.
# fit it to the trainset.
# transform the train and test set.
imputer = SimpleImputer(strategy='mean')
imputer.fit(x_train)
x_train = imputer.transform(x_train)
x_test_private = imputer.transform(x_test_private)

print("After Imputing:")
# Print number of empty cells both in x_train and x_test_private (1 line)
print("#Empty cells in x_train=", count_nans(x_train))
print("#Empty cells in x_test_private=", count_nans(x_test_private))



#Empty cells in x_train= 185698
#Empty cells in x_test_private= 19834
After Imputing:
#Empty cells in x_train= 0
#Empty cells in x_test_private= 0


### Features Scaling
1. Why we need to make all the features have same range of values?
<span style="color:blue"> This has something to do with some machine learning algorithms.
First, gradient-related algorithms (e.g., linear regression, logisitic regression, deep learning algorithms): feature scaling is important for the congergence of the algorithms as the gradient in one direction/feature could need more/long steps than the others and the scaling solves this issue and make the training stable and relatively faster.
Second, distance related algorithms (e.g., KNN, K-means). These algorithms depends on the distance between samples both in training and testing. Having features with different ranges would affect the distance measure and produce biased prediciton to the features with larger range.
for example: having two features (length in KM [10: 10000] and temperature in C° [20: 40]). The distance value will be much affected by the length feature.
2. What are the different strategies to scale the features?
<span style="color:blue">Min-Max scaling: transform all features to fall in this range: [0, 1].
$$x_i = \frac{x_i - min(\mathbf{x})}{max(\mathbf{x}) - min(\mathbf{x})}\\$$
Standard scaling: transform all features to have mean = 0 and standard deviation = 1. And by assuming that all features follow normal distribution, we can say that standard scaling turns featrues to be $\sim N(0,1)$.
$$x_i = \frac{x_i - mean(\mathbf{x})}{stdev(\mathbf{x})}\\$$
Robust scaling: robust to the outliers that may affect the previous methods in calculating their mean, stdev, min, or max. It uses the interquartile range to scale the features according to it.
Interquartile is the range between the 1st quartile and 3rd quartile.
$$x_i = \frac{x_i - Q_1(\mathbf{x})}{Q_3(\mathbf{x}) - Q_1(\mathbf{x})}\\$$
3. How to choose the strategy?
<span style="color:blue">Same as in imputation strategy, using a validation set or using cross-validation.

##### In the following cell, scale all the features with a scaler from your choice. Fit the scaler on the training set and transform both train and test set.

In [8]:
# Write your code here (5 lines):
# choose the scaler class.
# create the scaler object.
# fit it to the trainset.
# transform the train and test set.

from sklearn.preprocessing import RobustScaler

scaler = RobustScaler()

scaler.fit(x_train)
x_train = scaler.transform(x_train)
x_test_private = scaler.transform(x_test_private)

### Trainset splitting
Issues that lead to trainset splitting:
1. How we will measure the performance of our model?
<span style="color:blue"> Measuring performance on the trainset is not a good idea intuitively, as to estimate the generalization error, we need to evaluate the model on a data that wasn't seen before.
So we need to split the dataset that we have to two parts trainset and testset.
2. How we will tune the hyperparamters? Can it be done with testset?
<span style="color:blue"> If we tuned the hyperparamters on the testset, this will be considered as Data leakage as the hyperparams were tuned on the set that represent the evaluation, so there's a high chance that it will overfit the testset and fails to generalize (i.e., get high test accuracy but fails in production for example).
So, it is a must to split the data into 3 parts:
(70%) trainset to train the params , (10%) validation set to tune the hyperparameters (it can be used for other stuff like early stopping (tbe)), and (20%) testset to evaluate the model finally.

Issues to think of:
1. Is is the best strategy to get the best estimate for the true risk/performance of the system?
<span style="color:blue"> There's a chance that the testset is biased. So, a better way to estimate the true generalization error is to use cross validation.
Cross validation is a method where the dataset, is divided into k equally sized folds then we train on k-1 folds and test on the left one and repeat this process k times for each fold and get k estimates, finally we average all of these k accuracies to represent the final accuracy.
2. (Advanced) How to choose between two models?
If you have two models one with test accuracy 90% and the other with 91% (averaged from cross valdiation), which one to choose?
<span style="color:blue"> The naive/fast approach is to use the model with higher accuracy. But what if these estimates are not significantly different (we got them by chance), then a hypothesis test needs to be done.
After testing each model for each fold from the k-folds we can use "Paired t-test" to test if they are significantly different or not.
<br />(more advanced way, which is used in industry: A/B testing, watch about it [here](https://www.youtube.com/watch?v=zFMgpxG-chM))

##### In the next task, you will split your data into 3 parts, train, test, and validation by ratios: 70%, 20%, and 10% respectively.

In [9]:
from sklearn.model_selection import train_test_split

# Write your code here (2 lines)
x_train, x_test, y_train, y_test = train_test_split(x_train, y_train, test_size=0.2)
x_train, x_val, y_train, y_val = train_test_split(x_train, y_train, test_size=1/8)

print(x_train.shape, x_test.shape, x_val.shape)

(2705, 259) (773, 259) (387, 259)


### Data Plotting
Visualization is key-issue to get some insights and intuition about your dataset.
How to plot 259 features (259-dimensional data)? what we can plot -at maximum- is 2d or 3d data.

Hint: We should reduce the dimension. Read this [article](https://towardsdatascience.com/dimensionality-reduction-ways-and-intuitions-1b5e97592d8e).

So, Let's use UMAP to reduce the dimension of this dataset to be 2D.

In [12]:
!pip install umap

In [6]:
from sklearn.decomposition import PCA, FactorAnalysis, FastICA
from sklearn.manifold import MDS, TSNE
from sklearn import UMAP

dim_reducer = UMAP(n_components=2)
x_train_reduced = dim_reducer.fit_transform(x_train)

ImportError: cannot import name 'UMAP' from 'sklearn' (C:\Users\Ahuratus\AppData\Local\Programs\Python\Python39\lib\site-packages\sklearn\__init__.py)

Let's plot the reduced dataset on a 2D plane.
Use matplotlib to make a scatter plot for the reduced dataset.

In [1]:
import matplotlib.pyplot as plt
%matplotlib inline

# Write code to plot the reduced dataset in a scatter plot.
plt.scatter(x_train_reduced[:, 0], x_train_reduced[:, 1], marker='.')
plt.show()

NameError: name 'x_train_reduced' is not defined